metadata-ingestion/src/datahub/ingestion/source/powerbi/proxy.py

```diff
@@ -0,0 +1,911 @@
+            LOGGER.warning(message)
+            LOGGER.warning(f"{Constant.WorkspaceId}={workspace_id}")
+            LOGGER.warning(f"{Constant.DatasetId}={dataset_id}")
+            raise ConnectionError(message)
+
+        response_dict = response.json()
+        LOGGER.debug("datasets = {}".format(response_dict))
+        # PowerBi Always return the webURL, in-case if it is None then setting complete webURL to None instead of
+        # None/details
+        return PowerBiAPI.PowerBIDataset(
+            id=response_dict.get("id"),
+            name=response_dict.get("name"),
+            webUrl="{}/details".format(response_dict.get("webUrl"))
+            if response_dict.get("webUrl") is not None
+            else None,
+            workspace_id=workspace_id,
+            tables=[],
+        )
+
+    def get_data_sources(
+        self, dataset: PowerBIDataset
+    ) -> Optional[Dict[str, "PowerBiAPI.DataSource"]]:
+        """
+        Fetch the data source from PowerBi for the given dataset
+        """
+
+        datasource_get_endpoint: str = PowerBiAPI.API_ENDPOINTS[Constant.DATASOURCE_GET]
+        # Replace place holders
+        datasource_get_endpoint = datasource_get_endpoint.format(
+            POWERBI_BASE_URL=PowerBiAPI.BASE_URL,
+            WORKSPACE_ID=dataset.workspace_id,
+            DATASET_ID=dataset.id,
+        )
+        # Hit PowerBi
+        LOGGER.info(f"Request to datasource URL={datasource_get_endpoint}")
+        response = requests.get(
+            datasource_get_endpoint,
+            headers={Constant.Authorization: self.get_access_token()},
+        )
+
+        # Check if we got response from PowerBi
+        if response.status_code != 200:
+            message: str = "Failed to fetch datasource from power-bi for"
+            LOGGER.warning(message)
+            LOGGER.warning("{}={}".format(Constant.WorkspaceId, dataset.workspace_id))
+            LOGGER.warning("{}={}".format(Constant.DatasetId, dataset.id))
+            LOGGER.warning("{}={}".format(Constant.HTTP_RESPONSE_TEXT, response.text))
+            LOGGER.warning(
+                "{}={}".format(Constant.HTTP_RESPONSE_STATUS_CODE, response.status_code)
+            )
+
+            raise ConnectionError(message)
+
+        res = response.json()
+        value = res["value"]
+        if len(value) == 0:
+            LOGGER.info(
+                f"datasource is not found for dataset {dataset.name}({dataset.id})"
+            )
+
+            return None
+
+        data_sources: Dict[str, "PowerBiAPI.DataSource"] = {}
+        LOGGER.debug("data-sources = {}".format(value))
+        for datasource_dict in value:
+            # Create datasource instance with basic detail available
+            datasource = PowerBiAPI.DataSource(
+                id=datasource_dict.get(
+                    "datasourceId"
+                ),  # datasourceId is not available in all cases
+                type=datasource_dict["datasourceType"],
```

Comment:
are these fields guaranteed to be present?

Reply:
Yes